File: app/src/main/java/org/mesonet/app/reflection/ForecastModelParser.java
package org.mesonet.app.reflection;


import org.mesonet.app.forecast.ForecastModel;
import org.mesonet.app.formulas.UnitConverter;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

public class ForecastModelParser
{
    private static ForecastModelParser sModelParser;



    public static ForecastModelParser GetInstance()
    {
        if(sModelParser == null)
            sModelParser = new ForecastModelParser();

        return sModelParser;
    }



    private ForecastModelParser(){}



    public List<ForecastModel> Parse(String inStrValues)
    {
        if(inStrValues == null || inStrValues.isEmpty())
            return null;

        List<ForecastModel> result = new ArrayList<>();

        String[] strForecast = inStrValues.split(",");

        for(int i = 0, fieldOffset = 0; (fieldOffset + 8) < strForecast.length; i++, fieldOffset += 8) {

            try {
                ForecastModel forecastModel = new ForecastModel();

                if( Character.isDigit(strForecast[fieldOffset].charAt(0))) {
                    boolean found = false;

                    for(int j = 0; j < strForecast[fieldOffset].length() && !found; j++) {
                        if(!Character.isDigit(strForecast[fieldOffset].charAt(j)) && strForecast[fieldOffset].charAt(j) != ' ') {
                            found = true;
                            forecastModel.mTime = strForecast[fieldOffset].substring(j);
                        }
                    }
                }
                else
                    forecastModel.mTime = strForecast[fieldOffset];
                forecastModel.mIconUrl = strForecast[fieldOffset + 1];
                forecastModel.mStatus = strForecast[fieldOffset + 2];
                forecastModel.mHighOrLow = ForecastModel.HighOrLow.valueOf(strForecast[fieldOffset + 3]);
                forecastModel.mTemp = Double.parseDouble(strForecast[fieldOffset + 4]);
//                forecastModel.mWindDirection = UnitConverter.CompassDirections.valueOf(strForecast[fieldOffset + 5]);
//                forecastModel.mWindGustsDirection = UnitConverter.CompassDirections.valueOf(strForecast[fieldOffset + 6]);
//                forecastModel.mWindSpd = Double.parseDouble(strForecast[fieldOffset + 7]);
//
//                if( Character.isDigit(strForecast[fieldOffset + 8].charAt(0))) {
//                    boolean found = false;
//
//                    for(int j = 0; j < strForecast[fieldOffset + 8].length() && !found; j++) {
//                        if(!Character.isDigit(strForecast[fieldOffset + 8].charAt(j)) && strForecast[fieldOffset + 8].charAt(j) != ' ') {
//                            found = true;
//                            forecastModel.mWindGusts = Double.parseDouble(strForecast[fieldOffset + 8].substring(0, j));
//                        }
//                    }
//                }
//                else
//                    forecastModel.mWindGusts = Double.parseDouble(strForecast[fieldOffset + 8]);

                result.add(forecastModel);
            }
            catch (ArrayIndexOutOfBoundsException e)
            {
                e.printStackTrace();
            }
        }

        return result;
    }
}

File: dataprocessing/src/main/java/org/mesonet/dataprocessing/reflection/ForecastModelParser.kt
package org.mesonet.dataprocessing.reflection


import org.mesonet.dataprocessing.site.forecast.ForecastModel
import java.util.ArrayList

import javax.inject.Inject

class ForecastModelParser @Inject
internal constructor() {


    internal fun Parse(inStrValues: String?): List<ForecastModel>? {
        if (inStrValues == null || inStrValues.isEmpty())
            return null

        val result = ArrayList<ForecastModel>()

        val strForecast = inStrValues.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        var i = 0
        var fieldOffset = 0
        while (fieldOffset + 8 < strForecast.size) {

            try {
                val forecastModel = ForecastModel()

                if (Character.isDigit(strForecast[fieldOffset][0])) {
                    var found = false

                    var j = 0
                    while (j < strForecast[fieldOffset].length && !found) {
                        if (!Character.isDigit(strForecast[fieldOffset][j]) && strForecast[fieldOffset][j] != ' ') {
                            found = true
                            forecastModel.mTime = strForecast[fieldOffset].substring(j)
                        }
                        j++
                    }
                } else
                    forecastModel.mTime = strForecast[fieldOffset]
                    forecastModel.mIconUrl = strForecast[fieldOffset + 1]
                    forecastModel.mStatus = strForecast[fieldOffset + 2]
                    forecastModel.mHighOrLow = ForecastModel.HighOrLow.valueOf(strForecast[fieldOffset + 3])
                    forecastModel.mTemp = java.lang.Double.parseDouble(strForecast[fieldOffset + 4])

                result.add(forecastModel)
            } catch (e: ArrayIndexOutOfBoundsException) {
                e.printStackTrace()
            }
            catch (e: NumberFormatException)
            {
                e.printStackTrace()
            }
            catch (e: IllegalArgumentException)
            {
                e.printStackTrace()
            }

            i++
            fieldOffset += 8
        }

        return result
    }
}
